File: integration-test/src/main/java/com/sequenceiq/it/spark/spi/CloudMetaDataStatuses.java
package com.sequenceiq.it.spark.spi;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Objects;
import java.util.Optional;

import com.google.common.reflect.TypeToken;
import com.google.gson.Gson;
import com.sequenceiq.cloudbreak.cloud.model.CloudInstance;
import com.sequenceiq.cloudbreak.cloud.model.CloudVmInstanceStatus;
import com.sequenceiq.cloudbreak.cloud.model.CloudVmMetaDataStatus;
import com.sequenceiq.cloudbreak.cloud.model.InstanceTemplate;
import com.sequenceiq.it.spark.ITResponse;

public class CloudMetaDataStatuses extends ITResponse {

    private final Map<String, CloudVmMetaDataStatus> instanceMap;

    public CloudMetaDataStatuses(Map<String, CloudVmMetaDataStatus> instanceMap) {
        this.instanceMap = instanceMap;
    }

    private List<CloudVmMetaDataStatus> createCloudVmMetaDataStatuses(List<CloudInstance> cloudInstances) {
        List<CloudVmMetaDataStatus> cloudVmMetaDataStatuses = new ArrayList<>();
        for (Entry<String, CloudVmMetaDataStatus> stringCloudVmMetaDataStatusEntry : instanceMap.entrySet()) {
            CloudVmMetaDataStatus oldCloudVmMetaDataStatus = stringCloudVmMetaDataStatusEntry.getValue();
            InstanceTemplate oldTemplate = oldCloudVmMetaDataStatus.getCloudVmInstanceStatus().getCloudInstance().getTemplate();
            Optional<CloudInstance> cloudInstance = cloudInstances.stream()
                    .filter(instance -> Objects.equals(instance.getTemplate().getPrivateId(), oldTemplate.getPrivateId())).findFirst();
            if (cloudInstance.isPresent()) {
                CloudInstance newCloudInstance = new CloudInstance(stringCloudVmMetaDataStatusEntry.getKey(), cloudInstance.get().getTemplate(),
                        cloudInstance.get().getParameters());
                CloudVmInstanceStatus cloudVmInstanceStatus = new CloudVmInstanceStatus(newCloudInstance,
                        oldCloudVmMetaDataStatus.getCloudVmInstanceStatus().getStatus());
                CloudVmMetaDataStatus newCloudVmMetaDataStatus = new CloudVmMetaDataStatus(cloudVmInstanceStatus, oldCloudVmMetaDataStatus.getMetaData());
                cloudVmMetaDataStatuses.add(newCloudVmMetaDataStatus);
            }
        }
        return cloudVmMetaDataStatuses;
    }

    @Override
    public Object handle(spark.Request request, spark.Response response) throws Exception {
        List<CloudInstance> cloudInstances = new Gson().fromJson(request.body(), new TypeToken<List<CloudInstance>>() {
        }.getType());
        return createCloudVmMetaDataStatuses(cloudInstances);
    }
}

File: integration-test/src/main/java/com/sequenceiq/it/spark/spi/CloudMetaDataStatuses.kt
package com.sequenceiq.it.spark.spi

import java.util.ArrayList

import org.slf4j.Logger
import org.slf4j.LoggerFactory

import com.google.common.reflect.TypeToken
import com.google.gson.Gson
import com.sequenceiq.cloudbreak.cloud.model.CloudInstance
import com.sequenceiq.cloudbreak.cloud.model.CloudInstanceMetaData
import com.sequenceiq.cloudbreak.cloud.model.CloudVmInstanceStatus
import com.sequenceiq.cloudbreak.cloud.model.CloudVmMetaDataStatus
import com.sequenceiq.cloudbreak.cloud.model.InstanceStatus
import com.sequenceiq.it.spark.ITResponse
import com.sequenceiq.it.util.ServerAddressGenerator

class CloudMetaDataStatuses(private val mockServerAddress: String, private val sshPort: Int) : ITResponse() {

    private fun createCloudVmMetaDataStatuses(cloudInstances: List<CloudInstance>): List<CloudVmMetaDataStatus> {
        val cloudVmMetaDataStatuses = ArrayList<CloudVmMetaDataStatus>()
        val numberOfServers = cloudInstances.size
        ServerAddressGenerator(numberOfServers).iterateOver { address, number ->
            val cloudInstance = cloudInstances[number]
            val cloudInstanceWithId = CloudInstance("instance-" + address, cloudInstance.template)
            val cloudVmInstanceStatus = CloudVmInstanceStatus(cloudInstanceWithId, InstanceStatus.STARTED)
            val cloudInstanceMetaData = CloudInstanceMetaData(address, mockServerAddress, sshPort, "MOCK")
            val cloudVmMetaDataStatus = CloudVmMetaDataStatus(cloudVmInstanceStatus, cloudInstanceMetaData)
            cloudVmMetaDataStatuses.add(cloudVmMetaDataStatus)
        }
        return cloudVmMetaDataStatuses
    }

    @Throws(Exception::class)
    override fun handle(request: spark.Request, response: spark.Response): Any {
        val cloudInstances = Gson().fromJson<List<CloudInstance>>(request.body(), object : TypeToken<List<CloudInstance>>() {

        }.type)
        return createCloudVmMetaDataStatuses(cloudInstances)
    }

    companion object {

        private val LOGGER = LoggerFactory.getLogger(CloudMetaDataStatuses::class.java)
    }
}
